### PR TITLE
experiment: isolate state within component

### PR DESCRIPTION
**Pull request by f0ssel**
Not a priority and this won't necessarily be merged,  but I think this reads cleaner and is easier to understand for me and I'd like to know what others think. 

Continuation of https://github.com/coder/coder/pull/4331 since I borked the branch. 

I'd like feedback from anyone with opinions, but especially want to here from the resident @coder/frontend peeps on how they feel about this change.

### CODE REVIEW

**Comment by BrunoQuaresma** [comment]
I think this use case would be better handled by a state machine. Too many states/effects to handle using useState and useEffect IMO

**Comment by BrunoQuaresma** [comment]
Interesting talk about `useEffect` https://www.youtube.com/watch?v=RW9TVhmxu6Q&ab_channel=ReactNext

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 1]
```diff
@@ -1,62 +1,155 @@
...
+  const [selectedTemplate, setSelectedTemplate] = useState<Template>()
+  const [templateSchema, setTemplateSchema] = useState<ParameterSchema[]>()
+  const [creatingWorkspace, setCreatingWorkspace] = useState<boolean>(false)
+  const [createWorkspaceError, setCreateWorkspaceError] = useState<Error | unknown>()
+  const [getTemplatesError, setGetTemplatesError] = useState<Error | unknown>()
+  const [getTemplateSchemaError, setGetTemplateSchemaError] = useState<Error | unknown>()
+  const [permissions, setPermissions] = useState<Record<string, boolean>>()
+  const [checkPermissionsError, setCheckPermissionsError] = useState<Error | unknown>()
+  const [workspaceQuota, setWorkspaceQuota] = useState<WorkspaceQuota>()
+  const [getWorkspaceQuotaError, setGetWorkspaceQuotaError] = useState<Error | unknown>()
```
Comment (Kira-Pilot) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:46]: I'm not sure if all of these need to be stateful. For example, we likely don't need a `useState` for `getTemplatesError`. Instead, we might have a generic stateful variable, `error`, that we hydrate with errors that come back from the API. 

To further simplify, you could look into [`useReducer`](https://reactjs.org/docs/hooks-reference.html#usereducer) or the context API. 

`useReducer` is like `useState` on steroids and is something I reach for if I find myself with a bunch of `useState`s.

The [context API](https://reactjs.org/docs/context.html#contextprovider) allows one to inject contexts (objects) to any component wrapped within a Provider.  In this case, you could lean on two contexts: `workspaceContext` and `templateContext`. Each might have properties like `workspace` or `template`, `loading`, and `error`.

File: site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx [hunk 2]
```diff
@@ -1,62 +1,155 @@
...
-    selectedTemplate,
-    getTemplateSchemaError,
-    getTemplatesError,
-    createWorkspaceError,
-    permissions,
-    workspaceQuota,
-    getWorkspaceQuotaError,
-    owner,
-  } = createWorkspaceState.context
+  useEffect(() => {
```
Comment (Kira-Pilot) [inline site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx:82]: Keep in mind that `useEffects` are really meant to handle side effects. It's best to use them sparingly and keep them pretty lean. I really like that you have multiple `useEffects` here broken up by concern - that's definitely the right way to think about them.

**Comment by Kira-Pilot** [comment]
This is a really neat rewrite - thanks for trying it out! Were we to get rid of `xstate` for this page, I think we'd want to use contexts so there aren't quite as many hooks. Hooks complicate the rendering of pages (contexts do, as well, but in a more encapsulated way). 
I do find it pretty telling that you found this easier to reason about (even with 10+ hooks) and I think we should talk about that as a team.
Curious to hear what my other FE teammates think.

**Comment by f0ssel** [comment]
Thanks for the reviews everyone, I think we should all do a chat sometime, I love hearing everyone's different opinions on this and it shows I have a lot to listen and learn from. 

> Interesting talk about `useEffect` https://www.youtube.com/watch?v=RW9TVhmxu6Q&ab_channel=ReactNext

@BrunoQuaresma I watched this video and I have some thoughts:
- The 2 render thing is interesting - do we have strict mode on? I'd like to know if this negatively affects this component. I think the answer here is a simple fetch component / utility like you have previously suggested to handle the debounce - I don't think this is really a state issue at the end of the day but something xstate does solve for us. 
- He explains a lot of cases where people misuse useEffect and don't expect as many re-renders - I read the criticism a lot but honestly don't think I run into it when I use them. The code here starts with 2 constants based on the URL, and waterfalls down with the correct amount of renders. Correct me if I'm wrong here - I don't have a great way to measure renders outside of watching the fetch calls in the network tab (which is all I've previously been concerned about). 
- I don't really buy the argument that "there is no room in the state model for effects" - I think this is only true if using something like redux like you mentioned. I believe react gives you the tools to make effective and performant components.
- I really liked his advice for having dependency arrays  that are as small as possible - I agree this is the goal with how the code should be structured. I think more work to keep components the right size and do a better job at encapsulating complexity will go a long way. 
- He ultimately says that `useEffect` is for synchronizing state - And I think that is ultimately what is happening in this component, right? We hydrate with empty values everywhere and as we get in data we render the parts that need it. 

I think the information is very interesting but idk if I follow his conclusions from the info presented - it sounds like the whole recommendation was based on the assumption that you already use a global state machine like redux.

**Comment by BrunoQuaresma** [comment]
@f0ssel nice you watched! My concern is less about performance and more about predictability. Using "state charts" - it could be any other framework, but I guess XState is the more stable one forJS - makes it easier to:
- Model the flow as a diagram. It allows you to visually see what is happening + play around with that before implementation
- Avoid bad transitions in an explicit way using guards and only allowing certain events to happen in certain states

My concern when using `useState` + `useEffect` to sync so many states is to know what states depend on each other and what actions cause transitions. Of course, that is not impossible but it can be tricky IMO.

**Comment by presleyp** [comment]
I hesitate to consider changing the fundamentals of our architecture after the RFC and so much code being written. But I'm happy to discuss the pros and cons of XState. I'll copy your points from before here so I can give you more specific thoughts:

> I like that the local state is all contained within the page component - still separate from view for storybook

I totally see the convenience of having it all in one place, but personally, I prefer the separation. React has always had trouble being able to separate logic from view for reuse and testing. Hooks are an improvement for sure, but they're still not first class objects, they have to be in a component. XState machines are [totally independent](https://xstate-catalogue.com/machines/pagination). The biggest win there I think is in testing, and in fairness we don't exploit that in v2 because unit testing machines didn't make it into our testing strategy. But I like simple components, I like that I can step through an XState machine in the VSCode simulation to look for bugs in my code or my actual understanding of the feature, and I think it's great that we can easily give each other feedback on the logical flow without necessarily processing all of the code (or having the component even written yet). 

> I wasn't sure how to do parallel fetching cleanly in xstate - this rewrite is technically faster since the data waterfall isn't completely linear anymore with the multiple useEffect.

You can handle parallelism in XState with a [parallel state node](https://xstate.js.org/docs/guides/parallel.html). We have a few; the [template](https://github.com/coder/coder/blob/main/site/src/xServices/template/templateXService.ts#L84) one is similar to this case, I think, pictured below. When you're in the `initialInfo` state, the `activeTemplatesVersion`, `templateResources`, `templateVersions`, and `templateDAUs` sub-machines are all live at the same time (the states with blue outlines are current states, and the events with blue backgrounds are currently possible). When they all finish by reaching their final states, the `initialInfo` state's `onDone` event is triggered. So in this machine, the first fetch happens alone, then there are three in parallel, and you can only move on when all three succeed. (I wish we handled errors here, but that's why I wrote all those error fixing tickets.)
Screen Shot 2022-10-06 at 10 44 32 PM


> I like using xstate for cross function state like auth and feature enablement
> I really like the boring idea of just fetching in useEffect at the appropriate layer and passing props down. It's verbose but explicit (shocking coming from a go programmer, I know).

It's so funny, because I think of XState as the verbose but explicit option. Message passing leaves a paper trail, the visualization clarifies, finite states make it clear/[enforce](https://stately.ai/blog/no-disabling-a-button-is-not-app-logic) which actions and substates are possible at a given time, the DSL kind of has one correct place for each piece of info. But that's how these things are; we come from different paradigms so different things seem obvious/mysterious to us. Definitely happy to chat more!

**Comment by f0ssel** [comment]
This was a really cool discussion, I'm glad I can see where the team is with this tooling and the level of buy in. I'm gonna close this in favor of leaning more into using `xstate` and continue the conversation as I learn more.

**Pull request closed without merging**